### PR TITLE
Add support for RA2A1

### PR DESCRIPTION
**Pull request by soburi**

### CODE REVIEW

**Comment by ydamigos** [comment]
@duynguyenxa @thaoluonguw  Could you have a look?

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#define BSP_HOCO_HZ             (24000000)
+#elif BSP_CFG_HOCO_FREQUENCY == 2
+                #define BSP_HOCO_HZ             (32000000)
+            #elif BSP_CFG_HOCO_FREQUENCY == 4
+                #define BSP_HOCO_HZ             (48000000)
+            #elif BSP_CFG_HOCO_FREQUENCY == 5
+                #define BSP_HOCO_HZ             (64000000)
+            #else
+                #error "Invalid HOCO frequency chosen (BSP_CFG_HOCO_FREQUENCY) in bsp_clock_cfg.h"
+            #endif
```
Comment (ydamigos) [inline zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h:24]: Please align them.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ ID Code
+ Note: To lock and disable the debug interface define BSP_ID_CODE_LOCKED in compiler settings.
+ WARNING: This will disable debug access to the part. However, ALeRASE command will be accepted, which will clear (reset) the ID code. After clearing ID code, debug access will be enabled.
+ */
+#if defined(BSP_ID_CODE_LOCKED)
+            #define BSP_CFG_ID_CODE_LONG_1 (0x00000000)
+            #define BSP_CFG_ID_CODE_LONG_2 (0x00000000)
+            #define BSP_CFG_ID_CODE_LONG_3 (0x00000000)
+            #define BSP_CFG_ID_CODE_LONG_4 (0x00000000)
+            #else
```
Comment (ydamigos) [inline zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h:72]: Please align them.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h (hunk 1 above)
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h]: Thank you for reviewing.
I formatted it with `clang-format`.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h (hunk 2 above)
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h]: Same as above.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#endif
+
+#define BSP_CFG_PLL_SOURCE                                                                         \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll), source, RA_PLL_SOURCE_DISABLE)
+#define BSP_CFG_PLL_DIV BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll), div, RA_PLL_DIV_1)
+#define BSP_CFG_PLL_MUL                                                                            \
+	DT_NODE_HAS_STATUS(DT_NODELABEL(pll), okay)                                                \
+	? BSP_CLOCKS_PLL_MUL(DT_PROP_BY_IDX(DT_NODELABEL(pll), mul, 0),                            \
+			     DT_PROP_BY_IDX(DT_NODELABEL(pll), mul, 1))                            \
+	: BSP_CLOCKS_PLL_MUL(0, 0)
```
Comment (duynguyenxa) [inline zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h:40]: @soburi , Could you please help to check commit https://github.com/zephyrproject-rtos/hal_renesas/pull/24/commits/4c4b6c711d3e38f00ace4f84610b8a937bf2966f and modify follow it, reason is this missing parenthesis "()" wrap around the macro so it make the multiplication value calculation "BSP_CLOCKS_PLL_MUL" incorrect

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+#define BSP_CFG_PLL1R_FREQUENCY_HZ BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll), freqr, 0)
+
+#define BSP_CFG_PLL2_SOURCE                                                                        \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll2), source, RA_PLL_SOURCE_DISABLE)
+#define BSP_CFG_PLL2_DIV BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll2), div, RA_PLL_DIV_1)
+#define BSP_CFG_PLL2_MUL                                                                           \
+	DT_NODE_HAS_STATUS(DT_NODELABEL(pll2), okay)                                               \
+	? BSP_CLOCKS_PLL_MUL(DT_PROP_BY_IDX(DT_NODELABEL(pll2), mul, 0),                           \
+			     DT_PROP_BY_IDX(DT_NODELABEL(pll2), mul, 1))                           \
+	: BSP_CLOCKS_PLL_MUL(0, 0)
```
Comment (duynguyenxa) [inline zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h:56]: @soburi , similar as https://github.com/zephyrproject-rtos/hal_renesas/pull/25/files#r1667938694

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h (hunk 3 above)
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h]: I modified it to follow 4c4b6c711d3e38f00ace4f84610b8a937bf2966f changes.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h (hunk 4 above)
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h]: Same as above.

**Review by ydamigos — APPROVED** [review]

**Comment by KhiemNguyenT** [comment]
@soburi and @ydamigos , let me check this. I'd like to align with other implementation by @thaoluonguw  and @duynguyenxa.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+#define BSP_CFG_SCICLK_DIV BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(sciclk), clk_div, 0)
+#define BSP_CFG_SPICLK_SOURCE                                                                      \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(spiclk), clk_src, RA_CLOCK_SOURCE_DISABLE)
+#define BSP_CFG_SPICLK_DIV BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(spiclk), clk_div, 0)
+#define BSP_CFG_ADCCLK_SOURCE                                                                      \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(adcclk), clk_src, RA_CLOCK_SOURCE_DISABLE)
+#define BSP_CFG_ADCCLK_DIV BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(adcclk), clk_div, 0)
+#define BSP_CFG_I3CCLK_SOURCE                                                                      \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(i3cclk), clk_src, RA_CLOCK_SOURCE_DISABLE)
+#define BSP_CFG_I3CCLK_DIV BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(i3cclk), clk_div, 0)
```
Comment (thaoluonguw) [inline zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h:123]: @soburi , RA2A1 didn't support for these clocks. Could you please update them match with supported clock which described in HWM?

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+#else
+#define BSP_CFG_PLL2_MUL BSP_CLOCKS_PLL_MUL(0, 0)
+#endif
+
+#define BSP_CFG_PL2ODIVP           BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll2), divp, RA_PLL_DIV_2)
+#define BSP_CFG_PLL2P_FREQUENCY_HZ BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll2), freqp, 0)
+#define BSP_CFG_PL2ODIVQ           BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll2), divq, RA_PLL_DIV_2)
+#define BSP_CFG_PLL2Q_FREQUENCY_HZ BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll2), freqq, 0)
+#define BSP_CFG_PL2ODIVR           BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll2), divr, RA_PLL_DIV_2)
+#define BSP_CFG_PLL2R_FREQUENCY_HZ BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll2), freqr, 0)
```
Comment (thaoluonguw) [inline zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h:69]: Hello @soburi , RA2A1 didn't support for these clocks. Could you please update them match with supported clock which described in HWM?
Please refer to https://github.com/zephyrproject-rtos/hal_renesas/blob/c2d795cfa138c85d928b69d259897279bae7b9da/zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h for what I mean.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h (hunk 5 above)
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h]: I was thinking that I could make it common, but as you pointed out, this should be a separate definition. I made it the definition created by the automatic generation of FSP.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h (hunk 6 above)
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h]: Same as above.

**Comment by soburi** [comment]
@duynguyenxa Could you revisit this?

**Comment by soburi** [comment]
@thaoluonguw @duynguyenxa Could you revisit this?

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pclkblock), sysclock_src,                   \
+					  RA_PLL_SOURCE_DISABLE)
+#define BSP_CFG_ICLK_DIV                                                                           \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(iclk), clk_div, RA_SYS_CLOCK_DIV_2)
+#define BSP_CFG_PCLKB_DIV                                                                          \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pclkb), clk_div, RA_SYS_CLOCK_DIV_8)
+#define BSP_CFG_PCLKD_DIV                                                                          \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pclkd), clk_div, RA_SYS_CLOCK_DIV_4)
+#define BSP_CFG_FCLK_DIV                                                                           \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(fclk), clk_div, RA_SYS_CLOCK_DIV_8)
```
Comment (thaoluonguw) [inline zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h:42]: Recommend using RA_SYS_CLOCK_DIV_2 as default configurations generated in e2studio.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+ * Copyright (c) 2020 - 2024 Renesas Electronics Corporation and/or its affiliates
+ * Copyright (c) 2024 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#ifndef BSP_MCU_FAMILY_CFG_H_
+#define BSP_MCU_FAMILY_CFG_H_
+#include "bsp_mcu_device_pn_cfg.h"
+#include "bsp_mcu_device_cfg.h"
+#include "bsp_override.h"
```
[the remark this answers is not in the document]
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_mcu_family_cfg.h]: Thank you for your comment.
Fixed it.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+#else
+#error "Invalid HOCO frequency, only can be set to 24MHz, 32MHz, 48MHz, 64MHz"
+#endif
+#define BSP_CFG_CLOCK_SOURCE                                                                       \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pclkblock), sysclock_src,                   \
+					  RA_PLL_SOURCE_DISABLE)
+#define BSP_CFG_ICLK_DIV                                                                           \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(iclk), clk_div, RA_SYS_CLOCK_DIV_2)
+#define BSP_CFG_PCLKB_DIV                                                                          \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pclkb), clk_div, RA_SYS_CLOCK_DIV_8)
```
[the remark this answers is not in the document]
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h]: Fixed.

ICLK -> / 1
PCLKB -> /2
PCLKD -> /1
FCLK -> 2/

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+#define BSP_CFG_PCLKD_DIV                                                                          \
+	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pclkd), clk_div, RA_SYS_CLOCK_DIV_4)
```
[the remark this answers is not in the document]
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h]: Fixed.

File: zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h (hunk 7 above)
Comment (soburi) [reply zephyr/ra/ra_cfg/fsp_cfg/bsp/ra2a1/bsp_clock_cfg.h]: Fixed.

**Review by thaoluonguw — APPROVED** [review]

**Review by duynguyenxa — APPROVED** [review]

**Review by KhiemNguyenT — APPROVED** [review]

**Pull request merged**